java/serve/src/main/java/io/ray/serve/api/ServeControllerClient.java

```diff
@@ -0,0 +1,410 @@
+                    .remote()
+                    .get()
+            : ((ActorHandle<ServeController>) controller)
+                .task(ServeController::getCheckpointPath)
+                .remote()
+                .get();
+  }
+
+  /**
+   * Retrieve RayServeHandle for service deployment to invoke it from Java.
+   *
+   * @param endpointName A registered service deployment.
+   * @param missingOk If true, then Serve won't check the endpoint is registered.
+   * @return
+   */
+  @SuppressWarnings("unchecked")
+  public RayServeHandle getHandle(String endpointName, boolean missingOk) {
+
+    String cacheKey = endpointName + "#" + missingOk;
+    if (handleCache.containsKey(cacheKey)) {
+      return handleCache.get(cacheKey);
+    }
+
+    Map<String, EndpointInfo> endpoints = null;
+    if (controller instanceof PyActorHandle) {
+      endpoints =
+          ServeProtoUtil.parseEndpointSet(
+              (byte[])
+                  ((PyActorHandle) controller)
+                      .task(PyActorMethod.of("get_all_endpoints"))
+                      .remote()
+                      .get()); // TODO-0528 endpoint list
+    } else {
+      LOGGER.warn("Client only support Python controller now.");
+      endpoints =
+          ServeProtoUtil.parseEndpointSet(
+              ((ActorHandle<? extends ServeController>) controller)
+                  .task(ServeController::getAllEndpoints)
+                  .remote()
+                  .get());
+    }
+
+    if (!missingOk && (endpoints == null || !endpoints.containsKey(endpointName))) {
+      throw new RayServeException(LogUtil.format("Endpoint {} does not exist.", endpointName));
+    }
+
+    RayServeHandle handle = new RayServeHandle(controller, endpointName, null, null);
+    handleCache.put(cacheKey, handle);
+    return handle;
+  }
+
+  public void deploy(
+      String name,
+      String deploymentDef,
+      Object[] initArgs,
+      Map<String, Object> rayActorOptions,
+      DeploymentConfig deploymentConfig,
+      String version,
+      String prevVersion,
+      String routePrefix,
+      String url,
+      Boolean blocking) {
+
+    if (deploymentConfig == null) {
+      deploymentConfig = new DeploymentConfig();
+    }
+    if (rayActorOptions == null) {
+      rayActorOptions = new HashMap<>();
+    }
+    // TODO set runtime_env to rayActorOptions is not supported now.
+    ReplicaConfig replicaConfig = new ReplicaConfig(deploymentDef, initArgs, rayActorOptions);
+
+    deploymentConfig.setVersion(version);
+    deploymentConfig.setPrevVersion(prevVersion);
+
+    if (deploymentConfig.getAutoscalingConfig() != null
+        && deploymentConfig.getMaxConcurrentQueries()
+            < deploymentConfig.getAutoscalingConfig().getTargetNumOngoingRequestsPerReplica()) {
+      LOGGER.warn(
+          "Autoscaling will never happen, because 'max_concurrent_queries' is less than 'target_num_ongoing_requests_per_replica' now.");
```

Comment:
```suggestion
          "Autoscaling will never happen, because 'max_concurrent_queries' is less than 'target_num_ongoing_requests_per_replica'.");
```
```suggestion
          "Autoscaling will never happen, because 'max_concurrent_queries' is less than 'target_num_ongoing_requests_per_replica' now.");
```

Reply:
Resolved.